File: trace_car/controller_node.py
#!/usr/bin/env python3
#conding: utf-8

import rclpy
from rclpy.node import Node
from trace_car.msg import Perception, CarInfo, GuiControl
import numpy as np
from std_msgs.msg import Float64MultiArray
from controller_core import ControllerCore
from sensor_msgs.msg import JointState
import argparse


TIMER_PERIOD = 0.01 # seconds, 100 hz

class ControllerNode(Node):

    def __init__(self, gui=False):
        super().__init__("controller_node")
        self.gui = gui
        self.ControllerCore = ControllerCore(self)
        self.timer = self.create_timer(TIMER_PERIOD, self.timer_callback)

        # subscribe   20 hz from ccd node publisher
        # receive the perception result
        # call back run the steer control
        self.create_subscription(
            Perception,
            '/perception',
            self.perception_callback,
            10
        )

        # subscribe from the joint states
        # receive the current speed
        # call back for speed control
        self.create_subscription(
            JointState,
            '/joint_states',
            self.joint_state_callback,
            10
        )

        # publisher for control command to car
        self.wheel_pub = self.create_publisher(
            Float64MultiArray,
            '/wheel_velocity_controller/commands',
            10
        )
        self.steer_pub = self.create_publisher(
            Float64MultiArray,
            '/steer_position_controller/commands',
            10
        )

        self.car_info_msg = CarInfo()  # for GUI display
        self.car_info_pub = self.create_publisher(CarInfo, "/car_info", 10)

        # if use GUI, publish ccd detection result to gui
        if self.gui:
            self.system_paused = True
            self.create_subscription(
                GuiControl,
                "/gui_control",
                self.gui_control_callback,
                5
            )

            self.get_logger().info("Controller Node started [GUI mode]")
        else:
            self.system_paused = False
            self.get_logger().info("Controller Node started")

    def on_reset(self):
        self.ControllerCore.on_reset()
        self.send_steer(0.0, 0.0)
        self.send_wheel(0.0, 0.0)


    def send_wheel(self, left: float, right: float):
        msg = Float64MultiArray()
        msg.data = [left, right]
        self.wheel_pub.publish(msg)

    def send_steer(self, left_angle: float, right_angle: float):
        msg = Float64MultiArray()
        msg.data = [left_angle, right_angle]
        self.steer_pub.publish(msg)

    def timer_callback(self):
        if self.system_paused:
            return

        # step to start the system, wait for a few frames to start
        if not self.ControllerCore.has_started:
            self.ControllerCore.start_step()
        else:
            if self.gui and self.ControllerCore.received_road_type is not None:
                # publish ccd detection result to GUI
                self.car_info_msg.target_steer = self.ControllerCore.target_steer
                self.car_info_msg.target_speed = self.ControllerCore.target_speed
                self.car_info_msg.steer_left = self.ControllerCore.steer_left
                self.car_info_msg.steer_right = self.ControllerCore.steer_right
                self.car_info_msg.speed_left = self.ControllerCore.speed_left
                self.car_info_msg.speed_right = self.ControllerCore.speed_right

                self.car_info_msg.output_steer = self.ControllerCore.output_steer
                self.car_info_msg.output_speed = self.ControllerCore.output_speed

                self.car_info_msg.road_type = self.ControllerCore.get_road_type()
                self.car_info_pub.publish(self.car_info_msg)

    def perception_callback(self, msg):
        if self.system_paused:
            return

        # 100 hz from ccd node publisher
        if not self.ControllerCore.has_started:
            self.send_steer(0.0, 0.0)
            return

        # steer control logic
        road_type = msg.road_type
        center_error = msg.center_error

        steer_L, steer_R = self.ControllerCore.steer_control_logic(road_type, center_error)
        self.send_steer(steer_L, steer_R)

    def joint_state_callback(self, msg):
        if self.system_paused:
            return

        # 100 hz from gazebo joint state publisher
        if not self.ControllerCore.has_started:
            self.send_wheel(0.0, 0.0)
            return

        # speed control logic
        joint_map_pos = dict(zip(msg.name, msg.position))

        # can used for how far the car has gone
        pos_wheel_L = joint_map_pos['wheel_L_back_to_body']
        pos_wheel_R = joint_map_pos['wheel_R_back_to_body']

        wheel_L, wheel_R = self.ControllerCore.wheel_control_logic(pos_wheel_L, pos_wheel_R)
        self.send_wheel(wheel_L, wheel_R)

    def gui_control_callback(self, msg):
        # receive control command from GUI
        if msg.command == GuiControl.START:
            self.get_logger().info("[CMD] STARTED")
            self.system_paused = False

        elif msg.command == GuiControl.PAUSE:
            self.get_logger().info("[CMD] PAUSED")
            self.system_paused = True
        elif msg.command == GuiControl.RESET:
            self.get_logger().info("[CMD] RESETTING – car returning to spawn …")
            self.system_paused = True
            self.on_reset()


# =========================
# main
# =========================
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui",
                        action="store_true",
                        help="Whether to use GUI control")
    args, _ = parser.parse_known_args()

    rclpy.init()

    node = ControllerNode(args.gui)

    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/controller_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float64_multi_array.hpp>
#include <sensor_msgs/msg/joint_state.hpp>

#include "trace_car/msg/perception.hpp"
#include "trace_car/msg/car_info.hpp"
#include "trace_car/controller_core.hpp"
#include "trace_car/msg/gui_control.hpp"

constexpr double CTRL_TIMER_PERIOD = 0.01;  // 100 Hz

class ControllerNode : public rclcpp::Node{
public:
    ControllerNode(bool gui)
    : Node("controller_node")
    {
        gui_ = gui;
        controller_core_ = std::make_unique<ControllerCore>(this);

        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(CTRL_TIMER_PERIOD),
            std::bind(&ControllerNode::timer_callback, this)
        );

        // subscriber for perception
        perception_sub_ = this->create_subscription<trace_car::msg::Perception>(
            "/perception",
            10,
            std::bind(&ControllerNode::perception_callback, this, std::placeholders::_1)
        );

        joint_states_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
            "/joint_states",
            10,
            std::bind(&ControllerNode::joint_states_callback, this, std::placeholders::_1)
        );

        // publisher for steer and wheel commands
        steer_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(
            "/steer_position_controller/commands",
            10
        );
        wheel_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(
            "/wheel_velocity_controller/commands",
            10
        );


        car_info_pub = this->create_publisher<trace_car::msg::CarInfo>(
            "/car_info",
            10
        );

        if (gui_) {
            system_paused_ = true;
            gui_control_sub_ = this->create_subscription<trace_car::msg::GuiControl>(
                "/gui_control",
                5,
                std::bind(&ControllerNode::gui_control_callback, this, std::placeholders::_1)
            );

            RCLCPP_INFO(this->get_logger(), "Controller Node started [GUI mode].");
        } else {
            RCLCPP_INFO(this->get_logger(), "Controller Node started.");
        }
    }

private:
    void on_reset() {
        controller_core_->on_reset();
        send_steer(0.0, 0.0);
        send_wheel(0.0, 0.0);
    }

    void send_wheel(double left, double right)
    {
        std_msgs::msg::Float64MultiArray msg;
        msg.data = {left, right};
        wheel_pub_->publish(msg);
    }

    void send_steer(double left_angle, double right_angle)
    {
        std_msgs::msg::Float64MultiArray msg;
        msg.data = {left_angle, right_angle};
        steer_pub_->publish(msg);
    }

    void timer_callback(){
        if (system_paused_) return;

        if (!controller_core_->has_started) {
            controller_core_->start_step();
        } else {
            // publish car info for GUI display
            if (controller_core_->get_road_type() >= 0) {
                trace_car::msg::CarInfo car_info_msg;

                car_info_msg.target_steer = controller_core_->target_steer;
                car_info_msg.target_speed = controller_core_->target_speed;
                car_info_msg.steer_left   = controller_core_->steer_L;
                car_info_msg.steer_right  = controller_core_->steer_R;
                car_info_msg.speed_left   = controller_core_->speed_L;
                car_info_msg.speed_right  = controller_core_->speed_R;

                car_info_msg.output_steer = controller_core_->output_steer;
                car_info_msg.output_speed = controller_core_->output_speed;

                car_info_msg.road_type = controller_core_->get_road_type();
                car_info_pub->publish(car_info_msg);
            }
        }
    }

    void perception_callback(const trace_car::msg::Perception::SharedPtr msg)
    {
        if (system_paused_) return;

        if (!controller_core_->has_started) {
            send_steer(0.0, 0.0);
            return;
        }

        // 100 hz from ccd node publisher
        int road_type = msg->road_type;
        int center_error = msg->center_error;

        // steer control
        auto [steer_L, steer_R] = controller_core_->steer_control_logic(road_type, center_error);
        send_steer(steer_L, steer_R);
    }

    void joint_states_callback(const sensor_msgs::msg::JointState::SharedPtr msg)
    {
        if (system_paused_) return;

        // 100 hz from gazebo joint state publisher
        if (!controller_core_->has_started) {
            send_wheel(0.0, 0.0);
            return;
        }

        double wheel_pos_L_ = 0.0;
        double wheel_pos_R_ = 0.0;

        for (size_t i = 0; i < msg->name.size(); i++)
        {
            if (msg->name[i] == "wheel_L_back_to_body")
                wheel_pos_L_ = msg->position[i];
            else if (msg->name[i] == "wheel_R_back_to_body")
                wheel_pos_R_ = msg->position[i];
        }

        // wheel speed control
        auto [wheel_L, wheel_R] = controller_core_->wheel_control_logic(wheel_pos_L_, wheel_pos_R_);
        send_wheel(wheel_L, wheel_R);
    }

    void gui_control_callback(const trace_car::msg::GuiControl::SharedPtr msg)
    {
        switch (msg->command) {
            case trace_car::msg::GuiControl::START:
                system_paused_ = false;
                break;
            case trace_car::msg::GuiControl::PAUSE:
                system_paused_ = true;
                break;
            case trace_car::msg::GuiControl::RESET:
                system_paused_ = true;
                on_reset();
                break;

        }
    }

    bool gui_ = false;
    bool system_paused_ = false;
    std::unique_ptr<ControllerCore> controller_core_;

    std::shared_ptr<rclcpp::Subscription<trace_car::msg::Perception>> perception_sub_;
    std::shared_ptr<rclcpp::Subscription<sensor_msgs::msg::JointState>> joint_states_sub_;
    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float64MultiArray>> steer_pub_;
    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float64MultiArray>> wheel_pub_;

    // use for gui
    std::shared_ptr<rclcpp::Publisher<trace_car::msg::CarInfo>> car_info_pub;
    std::shared_ptr<rclcpp::Subscription<trace_car::msg::GuiControl>> gui_control_sub_;


    rclcpp::TimerBase::SharedPtr timer_;
};

// ---- main ----
int main(int argc, char *argv[])
{
    bool use_gui = false;
    for(int i = 1; i < argc; i++) {
        std::string arg(argv[i]);
        if (arg == "--gui") {
            use_gui = true;
        }
    }

    rclcpp::init(argc, argv);
    auto node = std::make_shared<ControllerNode>(use_gui);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
